### PR TITLE
Add trusted hops parameters support

### PR DESCRIPTION
**Pull request by thomas-tacquet**
# Changes

- :gift: Add support for trusted hops parameter

/kind enhancement

**Release Note**

```release-note
- add support for trusted hops count parameter
```

**Docs**

```docs
This alow users to have trusted hops support
```

### CODE REVIEW

**Comment by thomas-tacquet** [comment]
@nak3 can you check this please ? 🙏

**Comment by nak3** [comment]
Thank you! I will take a look after back from the holiday on this Thursday.

**Review by nak3 — COMMENTED** [review]
Do you think there is any way to add an e2e test for this? I think we cannot add it simply but please let me double check with you.

**Comment by thomas-tacquet** [comment]
Hello @nak3  🧇 

About testing trusted hops support, in my environment on my cluster I validated trusted hops (0, 1, 2+...) using different layers of proxies to ensure the parameter is read and appliyed.

But in e2e tests I did not succed building such a way of testing 😬 I tried to find how other project implemented trusted hops testing but I did not find anything relevant that could help me.

Units tests in this PR ensure the parameter is read but it does not guarantee everyone that trusted hops works as expected 🤔 

I think we can merge this because I'm confident with this parameter adding and adding e2e tests for this can be done later ?

**Comment by thomas-tacquet** [comment]
Hello @nak3 👋 

Is this PR acceptable with current testing coverage ? Unit testing is ok but I can't figure out how to properly integrade good E2E tests for this case 🤔 

(IMHO ensuring the parameter is correctly set is "enough" as this PR scope is to add the parameter and not to implement the trusted hops logic.)

**Comment by nak3** [comment]
Sorry for the delay.
The PR looks good to me but I think we can add one tiny unit test in https://github.com/knative-sandbox/net-kourier/blob/main/pkg/envoy/api/http_connection_manager_test.go ?

File: config/200-config.yaml [hunk 1]
```diff
@@ -80,3 +80,7 @@ data:
     # NOTE THAT THIS IS AN EXPERIMENTAL / ALPHA FEATURE.
     enable-cryptomb: "false"
 
+
+    # trusted-hops-count Configure the number of additional ingress proxy hops from the
```
Comment (nak3) [inline config/200-config.yaml:84]:
```suggestion
    # Configures the number of additional ingress proxy hops from the
```

We should align the format with the other values.

File: pkg/config/configmap.go [hunk 2]
```diff
@@ -108,6 +110,9 @@ type Kourier struct {
 	IdleTimeout time.Duration
 	// Desire level of incoming traffic isolation
 	TrafficIsolation TrafficIsolationType
+	// TrustedHopsCount Configure the number of additional ingress proxy hops from the
```
Comment (nak3) [inline pkg/config/configmap.go:113]:
```suggestion
	// TrustedHopsCount configures the number of additional ingress proxy hops from the
```

nit

**Comment by thomas-tacquet** [comment]
No problem @nak3 😁 
I added an unit test to check for different values (0, 1 and max 👌)

File: pkg/envoy/api/http_connection_manager_test.go [hunk 3]
```diff
@@ -110,3 +111,40 @@ func TestNewRouteConfig(t *testing.T) {
 
 	assert.DeepEqual(t, got, want, protocmp.Transform())
 }
+
+func TestNewHTTPConnectionManagerWithoutTrustedHopsDisabled(t *testing.T) {
```
Comment (nak3) [inline pkg/envoy/api/http_connection_manager_test.go:115]:
```suggestion
func TestNewHTTPConnectionManagerWithTrustedHops(t *testing.T) {
```
Your test name is also fine so please feel free to apply the suggestion or not by your preference.

**Comment by nak3** [comment]
/lgtm
/approve

(I think I don't have the approve permission now...)

File: pkg/envoy/api/http_connection_manager_test.go (hunk 3 above)
Comment (thomas-tacquet) [reply pkg/envoy/api/http_connection_manager_test.go]: Nice spot, at first I wanted to create different test and then moved to a single loop test 😁

**Review by evankanderson — COMMENTED** [review]
/lgtm
/approve

If you wanted to write an integration test, I would simply have a client send a request with 2-3 `x-forwarded-for` levels of headers, and have the receiving program echo back the resulting headers in the request body.  The client could then determine if the correct set of headers had been accepted.

In any case, this seems reasonable, so approving.

**Comment by evankanderson** [comment]
@nak3 -- you should still be able to have approvers if you want it, IMO.  (Though we should check with @dprotaso and @psschwei )

Let me know if you want that bit -- we can add you either to `networking-approvers` or specifically for `net-kourier`.

**Comment by nak3** [comment]
Thank you! Yes, I noticed that it was just dropped (by Dave's comment) and added myself to the approver yesterday https://github.com/knative/community/pull/1249

**Pull request merged**